components/script/dom/elementinternals.rs

```diff
@@ -0,0 +1,371 @@
+        }
+        Ok(self.is_instance_validatable())
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-form>
+    fn GetForm(&self) -> Fallible<Option<DomRoot<HTMLFormElement>>> {
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+        Ok(self.form_owner.get())
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-checkvalidity>
+    fn CheckValidity(&self) -> Fallible<bool> {
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+        Ok(self.check_validity())
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-reportvalidity>
+    fn ReportValidity(&self) -> Fallible<bool> {
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+        Ok(self.report_validity())
+    }
+}
+
+fn submission_value_from(value: &Option<FileOrUSVStringOrFormData>) -> SubmissionValue {
+    match value {
+        None => SubmissionValue::None,
+        Some(FileOrUSVStringOrFormData::File(f)) => SubmissionValue::File(DomRoot::from_ref(f)),
+        Some(FileOrUSVStringOrFormData::USVString(s)) => SubmissionValue::USVString(s.clone()),
+        Some(FileOrUSVStringOrFormData::FormData(fd)) => SubmissionValue::FormData(fd.datums()),
+    }
+}
+
+fn validity_bits(flags: &ValidityStateFlags) -> ValidationFlags {
+    let mut bits = ValidationFlags::empty();
+    if flags.valueMissing {
+        bits |= ValidationFlags::VALUE_MISSING;
+    }
+    if flags.typeMismatch {
+        bits |= ValidationFlags::TYPE_MISMATCH;
+    }
+    if flags.patternMismatch {
+        bits |= ValidationFlags::PATTERN_MISMATCH;
+    }
+    if flags.tooLong {
+        bits |= ValidationFlags::TOO_LONG;
+    }
+    if flags.tooShort {
+        bits |= ValidationFlags::TOO_SHORT;
+    }
+    if flags.rangeUnderflow {
+        bits |= ValidationFlags::RANGE_UNDERFLOW;
+    }
+    if flags.rangeOverflow {
+        bits |= ValidationFlags::RANGE_OVERFLOW;
+    }
+    if flags.stepMismatch {
+        bits |= ValidationFlags::STEP_MISMATCH;
+    }
+    if flags.badInput {
+        bits |= ValidationFlags::BAD_INPUT;
+    }
+    if flags.customError {
+        bits |= ValidationFlags::CUSTOM_ERROR;
+    }
+    bits
+}
+
+// Form-associated custom elements also need the Validatable trait.
+impl Validatable for ElementInternals {
+    fn as_element(&self) -> &Element {
+        debug_assert!(self.is_target_form_associated());
+        self.target_element.upcast::<Element>()
+    }
+
+    fn validity_state(&self) -> DomRoot<ValidityState> {
+        self.validity_state.or_init(|| {
+            ValidityState::new(
+                &window_from_node(self.target_element.upcast::<Node>()),
+                self.target_element.upcast(),
+            )
+        })
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#candidate-for-constraint-validation>
```

Comment:
This link might be broken because it has an extra slash? I think that https://html.spec.whatwg.org/multipage#candidate-for-constraint-validation works, but not this one.